hivemind/p2p/p2p_daemon.py

```diff
@@ -1,45 +1,188 @@
+            Multiaddr(f'/ip4/127.0.0.1/tcp/{self._client_listen_port}'))
 
-    def __init__(self, *args, **kwargs):
-        self._child = subprocess.Popen(args=self._make_process_args(args, kwargs))
+    async def _identify_client(self, delay):
+        await asyncio.sleep(delay)
         try:
-            stdout, stderr = self._child.communicate(timeout=0.2)
-        except subprocess.TimeoutExpired:
-            pass
-        else:
-            raise RuntimeError(f'p2p daemon exited with stderr: {stderr}')
+            encoded = await self._client.identify()
+            self.id = encoded[0].to_base58()
+        except Exception:
+            raise
 
-    def __enter__(self):
-        return self._child
+    def _assign_daemon_ports(self, host_port=None, daemon_listen_port=None):
+        self._host_port, self._daemon_listen_port = host_port, daemon_listen_port
+        if host_port is None:
+            self._host_port = find_open_port()
+            if daemon_listen_port is None:
+                self._daemon_listen_port = find_open_port()
+                while self._daemon_listen_port == self._host_port:
+                    self._daemon_listen_port = find_open_port()
+        if daemon_listen_port is None:
+            self._daemon_listen_port = find_open_port()
 
-    def __exit__(self, exc_type, exc_val, exc_tb):
-        self._kill_child()
+    @staticmethod
+    async def send_data(data, stream):
+        byte_str = pickle.dumps(data)
+        request = len(byte_str).to_bytes(P2P.HEADER_LEN, P2P.BYTEORDER) + byte_str
+        await stream.send_all(request)
+
+    @staticmethod
+    async def receive_data(stream, max_bytes=(1 << 16)):
+        data = await stream.receive_some(max_bytes)
+        content_length = int.from_bytes(data[:P2P.HEADER_LEN], P2P.BYTEORDER)
+        chunks = [data[P2P.HEADER_LEN:]]
+        curr_length = len(data) - P2P.HEADER_LEN
+
+        while curr_length < content_length:
+            data = await stream.receive_some(max_bytes)
+            chunks.append(data)
+            curr_length += len(data)
+
+        return pickle.loads(b''.join(chunks))
+
+    @staticmethod
+    def _handle_stream(handle):
+        async def do_handle_stream(stream_info, stream):
+            try:
+                request = await P2P.receive_data(stream)
+                result = handle(request)
+                await P2P.send_data(result, stream)
+            except Exception as exc:
+                await P2P.send_data(exc, stream)
+            finally:
+                await stream.close()
+
+        return do_handle_stream
+
+    def start_listening(self):
+        async def listen():
+            async with self._client.listen():
+                await self._listen_event.wait()
+
+        self._listen_task = asyncio.create_task(listen())
+
+    async def stop_listening(self):
+        if self._listen_task is not None:
+            self._listen_event.set()
```

Comment:
I suggest to rename `_listen_event` to `_stopping_event` for clarity.

Reply:
Done